### PR TITLE
Add Graph display style for atmos debug overlay

### PR DESCRIPTION
**Pull request by drcatdoctor**
## About the PR 
Adds a graph display style to atmos debug for when you want to see more at once.

```
> showatmos
> atvstyle Graph
```

`atvstyle` is a new command that swaps overlays. The old style (atvstyle Block) is the default and remains untouched.

The command `atvmode` exists and works like it always has, if you just want to use it as before, but when in Graph display style, it toggles things on and off instead of swapping modes. In Block display style it has the new feature of letting you toggle drawing of blocked air directions, pressure directions and excited groups. Using the command `atvmode` by itself gives you some hopefully self-explanatory help.

What atvstyle Graph does is try to divide the tile into four quadrants and use them for various info but still let you see what's going on in the game underneath:
```
flow |
dir  |
     |
-----------
notes|log
temp |bar
kPa  |graph
```

1) "flow dir": a big dumb arrow kind of thing. Glows pink when there is a PressureDirection and is grey when there's only a LastPressureDirection.

2) "notes": "EXC" for a tile part of an Excited Group and "EXC HOT" for one that is also a hotspot. (Also "HOT" for hot but not in an excited group but iirc that isn't supposed to be possible.) temp and kPa are self explanatory.

3) the "log bar graph" - not really useful to determine what the real numbers are, but it gives you a general handwavy sense of what gases are present and if there's an "important" amount of them. No bar appears if there's literally zero of a gas, but any amount above zero will draw at least a couple pixels.

**Screenshots**

With integer scaling
img

Without
img

I tried to record a video but my laptop is a potato, at least when it comes to ss14 + rider + recording at the same time, or I am using the wrong tool or something.

### CODE REVIEW

File: Content.Client/Atmos/EntitySystems/AtmosDebugOverlaySystem.cs [hunk 1]
```diff
@@ -42,8 +42,10 @@ public override void Initialize()
             SubscribeLocalEvent<GridRemovalEvent>(OnGridRemoved);
 
             var overlayManager = IoCManager.Resolve<IOverlayManager>();
-            if(!overlayManager.HasOverlay<AtmosDebugOverlay>())
-                overlayManager.AddOverlay(new AtmosDebugOverlay());
+            if(!overlayManager.HasOverlay<AtmosDebugWorldspaceOverlay>())
+                overlayManager.AddOverlay(new AtmosDebugWorldspaceOverlay());
+            if(!overlayManager.HasOverlay<AtmosDebugScreenspaceOverlay>())
+                overlayManager.AddOverlay(new AtmosDebugScreenspaceOverlay());
```
Comment (metalgearsloth) [inline Content.Client/Atmos/EntitySystems/AtmosDebugOverlaySystem.cs:48]: You can have multiple viewing spaces in one overlay so more is redundant, also the hasoverlay check isn't needed anymore as it doesn't throw.

File: Content.Client/Atmos/EntitySystems/AtmosDebugOverlaySystem.cs [hunk 2]
```diff
@@ -68,8 +70,10 @@ public override void Shutdown()
         {
             base.Shutdown();
             var overlayManager = IoCManager.Resolve<IOverlayManager>();
-            if (overlayManager.HasOverlay<AtmosDebugOverlay>())
-                overlayManager.RemoveOverlay<AtmosDebugOverlay>();
+            if (overlayManager.HasOverlay<AtmosDebugWorldspaceOverlay>())
+                overlayManager.RemoveOverlay<AtmosDebugWorldspaceOverlay>();
+            if(!overlayManager.HasOverlay<AtmosDebugScreenspaceOverlay>())
+                overlayManager.RemoveOverlay<AtmosDebugScreenspaceOverlay>();
```
Comment (metalgearsloth) [inline Content.Client/Atmos/EntitySystems/AtmosDebugOverlaySystem.cs:76]: Same here for removes

File: Content.Client/Atmos/EntitySystems/AtmosDebugOverlaySystem.cs [hunk 3]
```diff
@@ -99,6 +103,7 @@ internal enum AtmosDebugOverlayMode : byte
     {
         TotalMoles,
         GasMoles,
-        Temperature
+        Temperature,
+        Everything
```
Comment (metalgearsloth) [inline Content.Client/Atmos/EntitySystems/AtmosDebugOverlaySystem.cs:107]: Better off just making them flags.

Vera's most familiar with the atmosdebugoverlay anyway so she should review it.

File: Content.Client/Atmos/Overlays/AtmosDebugScreenspaceOverlay.cs [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+public sealed class AtmosDebugScreenspaceOverlay : Overlay
+{
+    private readonly AtmosDebugOverlaySystem _atmosDebugOverlaySystem;
+
+    [Dependency] private readonly IMapManager _mapManager = default!;
+    [Dependency] private readonly IEyeManager _eyeManager = default!;
+    [Dependency] private readonly IResourceCache _resourceCache = default!;
+    private readonly VectorFont _font;
+
+    public override OverlaySpace Space => OverlaySpace.ScreenSpace;
```
Comment (metalgearsloth) [inline Content.Client/Atmos/Overlays/AtmosDebugScreenspaceOverlay.cs:19]: As above you can use multiple spaces.

File: Content.Shared/Atmos/Atmospherics.cs [hunk 5]
```diff
@@ -271,6 +271,19 @@ static Atmospherics()
...
+        ///     Currently only used in the debug overlay.
+        /// </summary>
+        public static readonly Dictionary<Gas, Color> GasColors = new(){
+            { Gas.Oxygen, Color.Aqua },
+            { Gas.Nitrogen, Color.Green },
+            { Gas.CarbonDioxide, Color.DarkGray },
+            { Gas.Plasma, Color.HotPink },
+            { Gas.Tritium, Color.Lime },
+            { Gas.WaterVapor, Color.LightGray }
+        };
```
Comment (metalgearsloth) [inline Content.Shared/Atmos/Atmospherics.cs:286]: Just make it an array and comment it.

**Review by metalgearsloth — CHANGES_REQUESTED** [review]
@Zumorica

File: Content.Client/Atmos/EntitySystems/AtmosDebugOverlaySystem.cs (hunk 1 above)
Comment (drcatdoctor) [reply Content.Client/Atmos/EntitySystems/AtmosDebugOverlaySystem.cs:48]: Oh, cool, ok. Didn't notice you could do that. Thanks for the review.

File: Content.Client/Atmos/Overlays/AtmosDebugBlockOverlay.cs [hunk 6]
```diff
@@ -96,61 +96,71 @@ protected override void Draw(in OverlayDrawArgs args)
                             }
                             else if (pass == 1)
                             {
-                                // -- Blocked Directions --
-                                void CheckAndShowBlockDir(AtmosDirection dir)
```
Comment (drcatdoctor) [inline Content.Client/Atmos/Overlays/AtmosDebugBlockOverlay.cs:100]: github diff is being fussy, I just moved these in a tab level and added if flag checks

File: Content.Client/Atmos/EntitySystems/AtmosDebugOverlaySystem.cs [hunk 7]
```diff
@@ -1,49 +1,96 @@
...
+            AtmosDebugShowMode.TotalMoles | AtmosDebugShowMode.GasMoles | AtmosDebugShowMode.Temperature
+                | AtmosDebugShowMode.BlockDirections | AtmosDebugShowMode.ExcitedGroups
+                | AtmosDebugShowMode.FlowDirections
+        };
+
         public override void Initialize()
         {
             base.Initialize();
 
+            IoCManager.InjectDependencies(this);
```
Comment (mirrorcult) [inline Content.Client/Atmos/EntitySystems/AtmosDebugOverlaySystem.cs:67]: systems automatically have dependencies injected, this is not necessary

File: Content.Client/Atmos/Overlays/AtmosDebugGraphOverlay.cs [hunk 8]
```diff
@@ -0,0 +1,241 @@
...
+                var lineTop = lineBottom + new Vector2(0, linesMaxLen * interp);
+                drawHandle.DrawLine(lineBottom, lineTop, Atmospherics.GasDisplayColors[i]);
+            }
+        }
+    }
+
+    private static ValueTuple<Vector2, Vector2> TileFaceInDirection(TileRef tile, AtmosDirection dir)
+    {
+        // Account for South being 0.
+        var atmosAngle = dir.ToAngle() - Angle.FromDegrees(90);
```
Comment (mirrorcult) [inline Content.Client/Atmos/Overlays/AtmosDebugGraphOverlay.cs:111]: is this supposed to be + or -90?

```suggestion
        var atmosAngle = dir.ToWorldAngle();
```

File: Content.Shared/Atmos/Atmospherics.cs [hunk 9]
```diff
@@ -271,6 +271,21 @@ static Atmospherics()
...
 
         #endregion
+
+        /// <summary>
+        ///     Color to represent the gas in UIs and such.
+        ///     Currently only used in the debug overlay.
+        ///
+        ///     Ordering needs to be the same as <see cref="Gas"/>.
+        /// </summary>
+        public static readonly Color[] GasDisplayColors = {
```
Comment (mirrorcult) [inline Content.Shared/Atmos/Atmospherics.cs:281]: this needs to have a size of `Atmospherics.TotalNumberOfGases` (and probably some fallback color) so no out of bounds accesses if someone adds a new gas and forgets to update this

File: Content.Client/Atmos/Overlays/AtmosDebugGraphOverlay.cs [hunk 10]
```diff
@@ -0,0 +1,241 @@
...
+                if (data.Moles[i] == 0f) continue;
+                var lineBottom = new Vector2(lineZeroX + (0.05f * i), linesBottomY);
+                var interp = (float)
+                    (Math.Log10(data.Moles[i] - _atmosDebugOverlaySystem.CfgBase) /
+                     Math.Log10(topValueToShow - _atmosDebugOverlaySystem.CfgBase));
+                if (interp < 0f) interp = 0.1f;
+                var lineTop = lineBottom + new Vector2(0, linesMaxLen * interp);
+                drawHandle.DrawLine(lineBottom, lineTop, Atmospherics.GasDisplayColors[i]);
+            }
+        }
```
Comment (mirrorcult) [inline Content.Client/Atmos/Overlays/AtmosDebugGraphOverlay.cs:105]: why is this indented in another block?

File: Content.Client/Atmos/Overlays/AtmosDebugGraphOverlay.cs [hunk 11]
```diff
@@ -0,0 +1,241 @@
...
+            {
+                var dataMaybeNull = _atmosDebugOverlaySystem.GetData(mapGrid.Index, tile.GridIndices);
+                if (!dataMaybeNull.HasValue) continue;
+
+                if ((_atmosDebugOverlaySystem.CfgMode & AtmosDebugShowMode.BlockDirections) != 0)
+                    DrawBlockDirections(drawHandle, tile, dataMaybeNull.Value);
+                if ((_atmosDebugOverlaySystem.CfgMode & AtmosDebugShowMode.GasMoles) != 0)
+                    DrawGasMolesBarChart(drawHandle, tile, dataMaybeNull.Value);
+                if ((_atmosDebugOverlaySystem.CfgMode & AtmosDebugShowMode.FlowDirections) != 0)
+                    DrawFlowDirections(drawHandle, tile, dataMaybeNull.Value);
```
Comment (mirrorcult) [inline Content.Client/Atmos/Overlays/AtmosDebugGraphOverlay.cs:69]:
```suggestion

                if (_atmosDebugOverlaySystem.GetData(mapGrid.Index, tile.GridIndices) is not { } data) continue;

                if ((_atmosDebugOverlaySystem.CfgMode & AtmosDebugShowMode.BlockDirections) != 0)
                    DrawBlockDirections(drawHandle, tile, data);
                if ((_atmosDebugOverlaySystem.CfgMode & AtmosDebugShowMode.GasMoles) != 0)
                    DrawGasMolesBarChart(drawHandle, tile, data);
                if ((_atmosDebugOverlaySystem.CfgMode & AtmosDebugShowMode.FlowDirections) != 0)
                    DrawFlowDirections(drawHandle, tile, data);
```

believe this works to make it a bit cleaner

File: Content.Client/Atmos/Overlays/AtmosDebugGraphOverlay.cs (hunk 8 above)
Comment (mirrorcult) [reply Content.Client/Atmos/Overlays/AtmosDebugGraphOverlay.cs:111]: also this seems duplicated with some stuff in `BlockOverlay`, is there a way to dedup this at all

**Review by mirrorcult — CHANGES_REQUESTED** [review]

**Comment by metalgearsloth** [comment]
You coming back to this @drcatdoctor ?

**Pull request closed without merging**